File: 5-最长回文/longestPalindrome.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        def expend(s, left, right):
            n = len(s)

            while left >= 0 and right < n:
                if s[left] == s[right]:
                    left = left - 1
                    right = right + 1
                else:
                    break
            
            return s[left+1: right]
        

        ans = ''
        if len(s) == 1:
            return s
        for i in range(len(s)):
            a1 = expend(s, i, i)
            a2 = expend(s, i, i+1)
            if len(a1) > len(ans):
                ans = a1
            if len(a2) > len(ans):
                ans = a2

        return ans

        

File: 5-最长回文/longestPalindrome.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        vector<vector<int>> dp(n, vector<int>(n));
        int begin = 0;
        int len = 0;
        for (int l = 0; l < n; ++l) {
            for (int i = 0; i + l < n; ++i) {
                int j = i + l;
                if (l == 0) {
                    dp[i][j] = 1;
                } else if (l == 1) {
                    dp[i][j] = (s[i] == s[j]);
                } else {
                    dp[i][j] = (s[i] == s[j] && dp[i + 1][j - 1]);
                }
                if (dp[i][j] && l > len) {
                    //ans = s.substr(i, l + 1);
                    begin = i;
                    len = l;
                }
            }
        }
        return s.substr(begin, len+1);
    }
};
